src/service/referral.service.ts

```diff
@@ -55,6 +56,28 @@ class ReferralService {
 
     async createReferral(createReferralDto: CreateReferralDto): Promise<Referral> {
         const { name, email, experience, address, roleId, phone, openingId, referredById, resume } = createReferralDto;
+
+        
+        const referrals = await this.referralRepository.findReferralsByEmail(email)
+        let referralWithSameRole;
+        if (referrals){
+            const previousopeningId = await this.openingService.getOpeningById(openingId);
+            if (previousopeningId){
```

Comment:
I think we don't need to check for previousOpeningId, if a new opening for the same role is added, the candidate should not be able to apply for that as well.